packages/multichain/src/middlewares/MultichainSubscriptionManager.ts

```diff
@@ -0,0 +1,154 @@
+
+type SubscriptionNotificationEvent = {
+  jsonrpc: '2.0';
+  method: 'eth_subscription';
+  params: {
+    subscription: Hex;
+    result: unknown;
+  };
+};
+
+type SubscriptionKey = {
+  scope: ExternalScopeString;
+  origin: string;
+  tabId?: number;
+};
+type SubscriptionEntry = SubscriptionKey & {
+  subscriptionManager: SubscriptionManager;
+};
+
+type MultichainSubscriptionManagerOptions = {
+  findNetworkClientIdByChainId: NetworkController['findNetworkClientIdByChainId'];
+  getNetworkClientById: NetworkController['getNetworkClientById'];
+};
+
+export class MultichainSubscriptionManager extends SafeEventEmitter {
+  #findNetworkClientIdByChainId: NetworkController['findNetworkClientIdByChainId'];
+
+  #getNetworkClientById: NetworkController['getNetworkClientById'];
+
+  #subscriptions: SubscriptionEntry[] = [];
+
+  constructor(options: MultichainSubscriptionManagerOptions) {
+    super();
+    this.#findNetworkClientIdByChainId = options.findNetworkClientIdByChainId;
+    this.#getNetworkClientById = options.getNetworkClientById;
+  }
+
+  onNotification(
+    { scope, origin, tabId }: SubscriptionKey,
+    { method, params }: SubscriptionNotificationEvent,
+  ) {
+    this.emit('notification', origin, tabId, {
+      method: 'wallet_notify',
+      params: {
+        scope,
+        notification: { method, params },
+      },
+    });
+  }
+
+  #getSubscriptionEntry({
+    scope,
+    origin,
+    tabId,
+  }: SubscriptionKey): SubscriptionEntry | undefined {
+    return this.#subscriptions.find((subscriptionEntry) => {
+      return (
+        subscriptionEntry.scope === scope &&
+        subscriptionEntry.origin === origin &&
+        subscriptionEntry.tabId === tabId
+      );
+    });
+  }
+
+  #removeSubscriptionEntry({ scope, origin, tabId }: SubscriptionEntry) {
+    this.#subscriptions = this.#subscriptions.filter((subscriptionEntry) => {
+      return (
+        subscriptionEntry.scope !== scope ||
+        subscriptionEntry.origin !== origin ||
+        subscriptionEntry.tabId !== tabId
+      );
+    });
+  }
+
+  subscribe(subscriptionKey: SubscriptionKey) {
+    const subscriptionEntry = this.#getSubscriptionEntry(subscriptionKey);
+    if (subscriptionEntry) {
+      return subscriptionEntry.subscriptionManager;
+    }
+
+    const networkClientId = this.#findNetworkClientIdByChainId(
+      toHex(parseCaipChainId(subscriptionKey.scope as CaipChainId).reference),
+    );
+    const networkClient = this.#getNetworkClientById(networkClientId);
+    const subscriptionManager = createSubscriptionManager({
+      blockTracker: networkClient.blockTracker,
+      provider: networkClient.provider,
+    });
+
+    subscriptionManager.events.on(
+      'notification',
+      (message: SubscriptionNotificationEvent) => {
+        this.onNotification(subscriptionKey, message);
+      },
+    );
+
+    this.#subscriptions.push({
+      ...subscriptionKey,
+      subscriptionManager,
+    });
+
+    return subscriptionManager;
```

Comment:
may want to augment/override `subscriptionManager.middleware.destroy` to call `unsubscribeByOriginAndTabId` here

Reply:
See pipeline cleanup fn in `setupProviderConnectionCaip`